plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java

```diff
@@ -410,6 +411,14 @@ private TableToPartitionMapping getTableToPartitionMapping(ConnectorSession sess
         for (int i = 0; i < min(partitionColumns.size(), tableColumns.size()); i++) {
             HiveType tableType = tableColumns.get(i).getType();
             HiveType partitionType = partitionColumns.get(i).getType();
+
+            if (tableType.getCategory() == ObjectInspector.Category.UNION) {
+                tableType = unionToStruct(tableType, typeManager);
```

Comment:
The change is bigger than it looks:

Instead of changing the "canCoerce" method, which is used by the coordinator when sending the split. I early casted the data type for the TableToPartitionMapping(part of the Split's information)'s columnCoercions field. 
Therefore, when worker node inspect the split's partition information, start to read the split information assigned by the coordinator, workers already found the partition type is struct instead of union.(This is used when creating the pagesource)

However, casting when creating StructCoercer is still needed, if we think this is a good approach, I can double check and see if creating StructCoercer can use the information from the TableToPartitionMapping

Reply:
First of all:
Since union type only allows one type at a time, I am not sure dereferencing for the union column is supported in general. I tried to look for examples as how union type are used, most examples I saw select the whole union column. Can you find a example that is doing dereferencing for the union type? I want to know what subscript is being used..


I don't think eventually "everything is struct - when they come to the worker." is possible, since union is a component type in ORC:
`Compound types have children columns that hold the values for their sub-elements. For example, a struct column has one child column for each field of the struct. Lists always have a single child column for the element values and maps always have two child columns. Union columns have one child column for each of the variant`

ORC reader/writer need to know it is handling a union column.

Do you have a preference for implementing "getSupportedType" or keep casting and using isSupportedType? seems the change is small, to reduce scope we should continue use isSupportedType?

Reply:
union to struct cast is only casting we have when building TableToPartitionMapping

Reply:
> Since union type only allows one type at a time, I am not sure dereferencing for the union column is supported in general.

Ultimately to process union types - I'd expect the end user to check the tag and get the non-null value somehow. Or expose the N values as separate column. i.e. the user needs to look at what's inside for processing. And if this is something that only throws an error if pushdown happens but not otherwise - then that's a bug. 

> I don't think eventually "everything is struct - when they come to the worker."

I just meant this in terms of things around hive pagesource code handling table and partitions metadata. For ORC, we have a layer on top which is the UnionColumnReader.

Reply:
summary from our discussion: remove this casting and make sure the rest still works.

Reply:
Now I understand the reason why I previously implemented in this way:

It is due to supporting dereference for union type. Especially for functions getHiveTypeForDereferences() and getHiveDereferenceNames(). Previously, Trino do not support dereference for union type, as we can see from these two functions, there were checkArgument statement enforcing the hiveType being struct type. (If you know how Hive has a dereference for union type, please let me know)

However, since right now we are implementing read union as struct, naturally, this read should be support dereference. 

So we have 3 solutions:

1. disable dereference for union type, making sure when read union type, the HiveColumnProjectionInfo is always empty
  needs further investigation.

2. casting union to struct while dereferencing: 
this needs major interfaces changes for getHiveTypeForDereferences() and getHiveDereferenceNames() and along its calling chain. In order to pass in typeManagers, which is necessary for casting union to struct. 

3. early casting union to struct, and using its struct representation for dereferencing. 
This implementation. 

I don't like approach 2, since it changes too much for this corner case. For this corner case, our aim is to reduce impact radius and ensure correctness rather than efficiency.

Reply:
moving on, we should be clear that Trino's Hive library only supports dereference for struct type(for Orc, Parquet, and Avro).

Reply:
Summary: we concluded to take approach 4.

Reply:
And we won't be supporting this feature for AVRO type, due to Hive Avro's implementation using Table schema for deserialization.